### PR TITLE
Draft: Feat/mapping iterator

### PR DESCRIPTION
**Pull request by jeswr**
A fresh branch, that does not use transduction, as discussed in #59.

I think this resolves all readability problems @RubenVerborgh though I'm hesitant to say that for certain until I look with fresh eyes in the morning.

~Getting close (I think) to correct readable behaviour - I just need to get one more set of new tests that I just created working. The currently failing tests test the case where the upstream iterator becomes non-readable without closing.~

~I'll come back to this in a few hrs later this evening :)~

### CODE REVIEW

**Comment by jeswr** [comment]
Just added a quick performance check we have

Main:
20000000 elems 20 maps 49566.92061899975
20000000 elems 20 maps 20 filters 24873.365212999284
100_000 iterators each with 5 maps and 100 elements  8609.967392999679

This PR
20000000 elems 20 maps 3584.7544269990176
20000000 elems 20 maps 20 filters 3637.598729001358
100_000 iterators each with 5 maps and 100 elements 1951.0139819998294

File: asynciterator.ts [hunk 1]
```diff
@@ -531,7 +536,7 @@ export class AsyncIterator<T> extends EventEmitter {
     @returns {module:asynciterator.AsyncIterator} A new iterator with items in the given range
   */
   range(start: number, end: number): AsyncIterator<T> {
-    return this.transform({ offset: start, limit: Math.max(end - start + 1, 0) });
+    return this.skip(start).take(Math.max(end - start + 1, 0));
```
Comment (RubenVerborgh) [inline asynciterator.ts]: (I wonder if we want to leave this until we have a better `take`, but not important.)

File: asynciterator.ts [hunk 2]
```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
   }
 }
 
+/**
+ * A synchronous mapping function from one element to another.
+ * The iterator performing the mapping is passed as a second argument.
```
Comment (RubenVerborgh) [inline asynciterator.ts]:
```suggestion
 * A return value of `null` means that nothing should be emitted for a particular item.
```

File: asynciterator.ts [hunk 3]
```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
...
+    else {
+      this._source._destination = this;
+      this._source.on('end', destinationClose);
+      this._source.on('error', destinationEmitError);
+      this._source.on('readable', destinationSetReadable);
+      // If we are given a source that is already readable
+      // then we need to set the state of this iterable to readable
+      // also as there is no guarantee that the is no forthcoming
+      // readable event from the source
+      this.readable = this._source.readable;
```
Comment (RubenVerborgh) [inline asynciterator.ts:832]: Makes sense; can also set it to `true` should this cause trouble (it shouldn't).

File: asynciterator.ts [hunk 4]
```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
...
+
+  read(): D | null {
+    // Do not read the source if the current iterator is ended
+    if (this.done)
+      return null;
+
+    // Do not read from the source if it is ended
+    // TODO: Possibly remove this after https://github.com/RubenVerborgh/AsyncIterator/pull/45 is merged
+    // the requirement for this to be removed is that the 'end' emit occurs synchronously when the ENDED
+    // state is set, which does not currently occur if endAsync is called.
```
Comment (RubenVerborgh) [inline asynciterator.ts]: I'd probably still keep this in any case. We should not make too strong guarantees about the source being well-behaved.

File: asynciterator.ts [hunk 5]
```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
...
+    if (this._source.done) {
+      this.close();
+      return null;
+    }
+
+    let item: D | S | null;
```
Comment (RubenVerborgh) [inline asynciterator.ts]: I'd be conservative here and have `item: S | null` and `mapped: D | null`. Maybe even declare `const mapped` inside of the loop.

File: asynciterator.ts [hunk 6]
```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
...
+    while ((item = this._source.read()) !== null) {
+      if ((item = this._map(item)) !== null)
+        return item;
+    }
+
+    // This will set readable to false on the current iterator
+    // if the source is no longer readable
+    this.readable = this._source.readable;
```
Comment (RubenVerborgh) [inline asynciterator.ts]: Right. That's sensible actually.
I'd have been okay with next iteration of `read` triggering that result, but this optimization makes sense. In fact, if the source returns `null`, we _know_ that it is not readable so we can just set it to `false` here?

File: asynciterator.ts [hunk 7]
```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
...
+      return null;
+
+    // Do not read from the source if it is ended
+    // TODO: Possibly remove this after https://github.com/RubenVerborgh/AsyncIterator/pull/45 is merged
+    // the requirement for this to be removed is that the 'end' emit occurs synchronously when the ENDED
+    // state is set, which does not currently occur if endAsync is called.
+    if (this._source.done) {
+      this.close();
+      return null;
+    }
```
Comment (RubenVerborgh) [inline asynciterator.ts]:
```suggestion
    if (!this._source.readable) {
      this.readable = false;
      if (this._source.done)
        this.close();
      return null;
    }
```

^ We should be maximally well-behaved and not attempt to call `.read` on the source unless it advertises itself as potentially readable.

File: asynciterator.ts [hunk 8]
```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
...
+    // This will set readable to false on the current iterator
+    // if the source is no longer readable
+    this.readable = this._source.readable;
+    return null;
+  }
+
+  public close() {
+    if (this._destroySource)
+      this._source.destroy();
+    super.close();
```
Comment (RubenVerborgh) [inline asynciterator.ts]: I think the order should be reversed (first bring ourselves in a closed state, then our internal source). But cross-check with other classes.

File: asynciterator.ts [hunk 9]
```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
...
+  }
+
+  /* Cleans up the source iterator and ends. */
+  protected _end(destroy: boolean) {
+    this._source.removeListener('end', destinationClose);
+    this._source.removeListener('error', destinationEmitError);
+    this._source.removeListener('readable', destinationSetReadable);
+    delete this._source._destination;
+    if (this._destroySource)
+      this._source.destroy();
```
Comment (RubenVerborgh) [inline asynciterator.ts:863]: That's duplicated from `close`; can it be removed there then (and thus, can the entire `close` implementation be removed)?

File: perf/MappingIterator-perf.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
+import { ArrayIterator, range } from '../dist/asynciterator.js'
+
+for (let i = 0; i < 100_000_000; i++) {
+  // Warming up
```
Comment (RubenVerborgh) [inline perf/MappingIterator-perf.js]: Nothing's warming though 😅

File: perf/MappingIterator-perf.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+function perf(iterator, description) {
+  return new Promise((res) => {
+    for (let i = 0; i < 10_000_000; i++) {
+      // Warming up
```
Comment (RubenVerborgh) [inline perf/MappingIterator-perf.js]: Still no heat.

**Review by RubenVerborgh — COMMENTED** [review]
Yes, good stuff, @jeswr. We're absolutely on the right track.

In general, we need to be conservative/defensive when programming derived classes, and not assume too many behaviors of them. I should probably have a list of _allowed assumptions_, but it's not black and white. Briefly, let's not assume that `source` is not one of the implementations we've written, but rather a more unstable one. We have to assume its state and its `readable` are correct, but that's about it.

File: asynciterator.ts (hunk 7 above)
Comment (jeswr) [reply asynciterator.ts]: @RubenVerborgh Is it worth opening an issue to apply this idea to other iterators, for instance in [`_readAndTransform`](https://github.com/RubenVerborgh/AsyncIterator/blob/a536bca2dfe84697e1a095c10eb1fcce40f7182c/asynciterator.ts#L1185) and [`_readAndTransformSimple`](https://github.com/RubenVerborgh/AsyncIterator/blob/a536bca2dfe84697e1a095c10eb1fcce40f7182c/asynciterator.ts#L1330-L1339).

File: asynciterator.ts (hunk 3 above)
Comment (jeswr) [reply asynciterator.ts:832]: Setting it to `true` is fine in terms of correctness. In terms of performance, I think it is best to not be readable if the source is not readable as this avoids firing unnecessary `readable` events and in turn prevents any downstream iterators trying to read from this and getting `null` returned.

I think this may also allow a more marked perf. improvement when https://github.com/RubenVerborgh/AsyncIterator/pull/45 is re-addressed as that PR gets rid of a lot of unnecessary `read` calls and in many cases can just about halve the number of `.read()` called when taking unions over transformations.

File: asynciterator.ts (hunk 6 above)
Comment (jeswr) [reply asynciterator.ts]: > In fact, if the source returns null, we know that it is not readable

I'm cautious about making that assumption, especially since it doesn't really yield that much in terms of performance gains. For instance, what if someone chose to implement a `FilterIterator` where the `.read` method was

```
read() {
  / * ... readability checks */

  const item = this.source.read()
  return this.filter(item) ? item : null;
}
```
then the source would still be readable.

Comment (jeswr) [reply asynciterator.ts]: Moreover I can't find anything that makes that assumption explicitly in the [readable API](https://nodejs.org/api/stream.html#event-readable). It makes the statement that "If no data is available to be read, null is returned", but does not say that a `null` return implies it is not readable.

File: asynciterator.ts (hunk 9 above)
Comment (jeswr) [reply asynciterator.ts:863]: Ah yes it can, the only thing that changes is that the destruction happens a tick later.

This will become moot with https://github.com/RubenVerborgh/AsyncIterator/pull/45 anyway which removes the `endAsync` that results in this next-tick behavior.

File: asynciterator.ts (hunk 6 above)
Comment (jacoscaz) [reply asynciterator.ts]: I wouldn’t be surprised if a filter method like that were to subtly break things in some use cases. I do think there’s an implicit assumption that a stream that reads null is also a stream that is not readable at that point in time. A safer implementation would read until the source reads null or until the source reads an item that matches the filter (as per what happens in `MappingIterator`).

Comment (jacoscaz) [reply asynciterator.ts]: Also, consider this paragraph from https://nodejs.org/api/stream.html#readablereadsize :

> When reading a large file `.read()` may return `null`, having consumed all buffered content so far, but there is still more data to come not yet buffered. In this case a new 'readable' event will be emitted when there is more data in the buffer.

AFAIU, this implies that whenever we `read()` the value `null` we can stop reading and resume doing so upon a new `readable` event.

File: asynciterator.ts (hunk 9 above)
Comment (jeswr) [reply asynciterator.ts:863]: I've made this change. The only thing I should flag is that in https://github.com/RubenVerborgh/AsyncIterator/pull/75/commits/b34fb6e71647fd6c58afe0734271864d6220461a I had to change from waiting a tick (where tests were failing) to doing a `beforeEach` which waits for the end event on the iterator that is being closed, and then will wait another tick before testing for destruction.

I think this should be ok, but is probably something that we can make even safer in #45 as noted in the comments in that PR.

File: asynciterator.ts (hunk 6 above)
Comment (jeswr) [reply asynciterator.ts]: I've been convinced, change pushed

File: asynciterator.ts (hunk 7 above)
Comment (RubenVerborgh) [reply asynciterator.ts]: Note that `TransformIterator` and `SimpleTransformIterator` also accept regular Node.js `Readable` instances, which do not define `readable`. So the goal there is to also work with non-`AsyncIterator`s (see `wrap`).

We can have the same goal for `MappingIterator`, but that requires different tests.

**Comment by RubenVerborgh** [comment]
Thanks, @jeswr, looking really good. Amazing what two fresh pairs of eyes can do, innit? 😉
I will make some small stylistic changes the code here and there, and then assign back to you for final approval.
Then we merge and release a new minor (unless you want to have `take` / the `UntilIterator` in as well).

**Comment by jeswr** [comment]
I think that since this is close, clean it up and then we can release. After that is done we could put most of the logic into an abstract class that `MapIterator`, `UntilIterator` and, eventually `CompositeIterator` can share.

I have in mind that it will look something like this (which does have all tests current passing and is over 2x faster on the current performance tests):

```ts
/**
 * A synchronous mapping function from one element to another.
 * A return value of `null` means that nothing should be emitted for a particular item.
 */
export type MapFunction<S, D = S> = (item: S) => D | null;

/**
 An iterator that calls a synchronous operation
 on every item from its source iterator.
 @extends module:asynciterator.AsyncIterator
*/
export abstract class SynchronousTransformIterator<S, D = S> extends AsyncIterator<D> {
  private readonly _destroySource: boolean;
  protected _source: InternalSource<S>;

  /**
   * Applies the given mapping to the source iterator.
   */
  constructor(
    _source: AsyncIterator<S>,
    options: SourcedIteratorOptions = {}
  ) {
    super();
    this._source = _validateSource(_source);
    this._destroySource = options.destroySource !== false;

    if (_source.done) {
      this.close();
    }
    else {
      this._source._destination = this;
      this._source.on('end', destinationClose);
      this._source.on('error', destinationEmitError);
      this._source.on('readable', destinationSetReadable);
      // If we are given a source that is already readable
      // then we need to set the state of this iterable to readable
      // also as there is no guarantee that the is no forthcoming
      // readable event from the source
      this.readable = this._source.readable;
    }
  }

  protected abstract safeRead(): D | null;

  read(): D | null {
    // Do not read the source if the current iterator is ended
    if (this.done)
      return null;

    // A source should only be read from if readable is true
    if (!this._source.readable) {
      this.readable = false;
      if (this._source.done)
        this.close();
      return null;
    }

    const item = this.safeRead();

    if (item !== null) {
      return item;
    }

    // This will set readable to false on the current iterator
    // if the source is no longer readable
    this.readable = false;
    return null;
  }

  /* Cleans up the source iterator and ends. */
  protected _end(destroy: boolean) {
    this._source.removeListener('end', destinationClose);
    this._source.removeListener('error', destinationEmitError);
    this._source.removeListener('readable', destinationSetReadable);
    delete this._source._destination;
    if (this._destroySource)
      this._source.destroy();
    super._end(destroy);
  }
}

export class MappingIterator<S, D = S> extends SynchronousTransformIterator<S, D> {
  constructor(
    source: AsyncIterator<S>,
    private readonly _map: MapFunction<S, D>,
    options: SourcedIteratorOptions = {}
  ) {
    super(source, options);
  }

  safeRead() {
    let item: S | null;
    while ((item = this._source.read()) !== null) {
      const mapped = this._map(item);
      if (mapped !== null)
        return mapped;
    }
    return null;
  }

  map<K>(map: MapFunction<D, K>, self?: any): AsyncIterator<K> {
    return new CompositeMappingIterator(this._source, [this._map, bind(map, self)], this);
  }
}

export class CompositeMappingIterator<S, D = S> extends SynchronousTransformIterator<S, D> {
  constructor(
    private root: AsyncIterator<S>,
    private mappings: MapFunction<any, any>[] = [],
    source: AsyncIterator<any>,
    options: SourcedIteratorOptions = {},
  ) {
    super(source, options);
  }

  safeRead() {
    // TODO: See if this is actually necessary
    // A source should only be read from if readable is true
    if (!this.root.readable) {
      this.readable = false;
      // TODO: See if this should be here
      if (this.root.done)
        this.close();
      return null;
    }

    let mapped : any = null;
    while (mapped === null && (mapped = this.root.read()) !== null) {
      for (let i = 0; i < this.mappings.length; i++) {
        mapped = this.mappings[i](mapped);
        if (mapped === null)
          break;
      }
    }
    return mapped;
  }

  map<K>(map: MapFunction<D, K>, self?: any): AsyncIterator<K> {
    return new CompositeMappingIterator(this.root, [...this.mappings, bind(map, self)], this);
  }
}

export class UntilIterator<S> extends SynchronousTransformIterator<S> {
  constructor(
    source: AsyncIterator<S>,
    private limit: number,
    options: SourcedIteratorOptions = {}
  ) {
    super(source, options);
  }

  safeRead() {
    if (this.limit <= 0) {
      this.close();
      return null;
    }
    this.limit -= 1;
    return this._source.read();
  }
}
```

I'll open up a branch with a cleaned up, and more heavily tested version of the code above (minus the `CompositeMappingIterator`) once the current branch is merged. The `CompositeMappingIterator` can then be cleanly addressed later on.

File: test/MappingIterator-test.js [hunk 12]
```diff
@@ -0,0 +1,1026 @@
...
+import { EventEmitter } from 'events';
+import { expect } from 'chai';
+
+describe('MappingIterator', () => {
+  describe('The MappingIterator function', () => {
+    describe('the result when called with `new`', () => {
+      let instance;
+
+      before(() => {
+        // TODO: Re-enable this
```
Comment (RubenVerborgh) [inline test/MappingIterator-test.js:21]: @jeswr What's this one?

**Comment by RubenVerborgh** [comment]
@jeswr I have done a review and have arrived at the code you can now see in this PR.
Apart from an overall approval, 3 questions:
- Any concerns about the simplified `read` logic? https://github.com/RubenVerborgh/AsyncIterator/pull/75/commits/eeb47967333b8116c3124f1857c9cf677a9d6ad4#diff-8c1065e179bf8c72bbaa6ae735f3be78aac1e609456c0086be7d2b3c1dae66ecR844
- Any concerns about the`readable` logic? https://github.com/RubenVerborgh/AsyncIterator/pull/75/commits/eeb47967333b8116c3124f1857c9cf677a9d6ad4#diff-8c1065e179bf8c72bbaa6ae735f3be78aac1e609456c0086be7d2b3c1dae66ecL1334-L1337 The original was a bit naughty as superfluous events would be emitted; I'm just checking if there was a reason for that. The unit tests worked fine without it.
- Any concerns related to the changes and your proposed `SynchronousTransformIterator`?

**Comment by jeswr** [comment]
> Any concerns about the simplified read logic?

 - I usually call `this._source.read()` outside of `if` statements as this reduces the amount of branching required to get to the call which I assume to be the most common case when `.read()` is called. ~Would need to do perf. checks to see if this *actually* has any performance impact though.~ Looks like this doesn't make a difference in practise though

```
Original

[jeswr@orangejuly AsyncIterator]$ node perf/MappingIterator-perf.js 
20000000 elems 20 maps                                   1678.1333249993622
20000000 elems 20 maps 20 filter                         1854.2675620019436
100_000 iterators each with 5 maps and 100 elements      2297.956739999354
[jeswr@orangejuly AsyncIterator]$ node perf/MappingIterator-perf.js 
20000000 elems 20 maps                                   1611.3272369988263
20000000 elems 20 maps 20 filter                         1734.021374002099
100_000 iterators each with 5 maps and 100 elements      2286.2911899983883
[jeswr@orangejuly AsyncIterator]$ node perf/MappingIterator-perf.js 
20000000 elems 20 maps                                   1594.8407809995115
20000000 elems 20 maps 20 filter                         1748.050572000444
100_000 iterators each with 5 maps and 100 elements      2127.236763998866

---

Now

[jeswr@orangejuly AsyncIterator]$ node perf/MappingIterator-perf.js 
20000000 elems 20 maps                                   1703.3545530028641
20000000 elems 20 maps 20 filter                         1865.0185250006616
100_000 iterators each with 5 maps and 100 elements      2192.7923329994082
[jeswr@orangejuly AsyncIterator]$ node perf/MappingIterator-perf.js 
20000000 elems 20 maps                                   1602.2818830013275
20000000 elems 20 maps 20 filter                         1791.3512960001826
100_000 iterators each with 5 maps and 100 elements      2325.8517739996314
[jeswr@orangejuly AsyncIterator]$ node perf/MappingIterator-perf.js 
20000000 elems 20 maps                                   1611.5303689986467
20000000 elems 20 maps 20 filter                         1763.5889509990811
100_000 iterators each with 5 maps and 100 elements      2306.6571919992566

```
 
 - In terms of getting the most out of #45 once it is merged, I think the `.done` check should be an `if` afterwards rather than an `else if` https://github.com/RubenVerborgh/AsyncIterator/blob/eeb47967333b8116c3124f1857c9cf677a9d6ad4/asynciterator.ts#L854-L856 to allow chained iterators to close faster.

**Comment by jeswr** [comment]
> Any concerns about thereadable logic? https://github.com/RubenVerborgh/AsyncIterator/commit/eeb47967333b8116c3124f1857c9cf677a9d6ad4#diff-8c1065e179bf8c72bbaa6ae735f3be78aac1e609456c0086be7d2b3c1dae66ecL1334-L1337 The original was a bit naughty as superfluous events would be emitted; I'm just checking if there was a reason for that. The unit tests worked fine without it.

Ooh interesting - I thought those were failing before without. I think it was historical when @jacoscaz needed a similar pattern to get unit tests passing.

If tests are working then I think it is safe to get rid of it provided we have good enough coverage with asynchronous `source`s for the iterator.

**Comment by jeswr** [comment]
> Any concerns related to the changes and your proposed SynchronousTransformIterator?

Nope - tests still pass making the changes to convert to the `SynchronousTransformIterator` proposal.

~I don't think so - it should just be a matter of moving the `.safeRead()` call. I'll quickly run some tests to confirm.~

File: asynciterator.ts [hunk 13]
```diff
@@ -777,9 +791,88 @@ export class IntegerIterator extends AsyncIterator<number> {
...
+    if (!this.done) {
+      if (this._source.readable) {
+        // Try to read an item that maps to a non-null value
+        let item: S | null, mapped: D | null;
+        while ((item = this._source.read()) !== null) {
+          if ((mapped = this._map(item)) !== null)
+            return mapped;
+        }
+      }
+      else if (this._source.done) {
```
Comment (jeswr) [inline asynciterator.ts]:
```suggestion
      if (this._source.done) {
```

Only one suggested change - see https://github.com/RubenVerborgh/AsyncIterator/pull/75#issuecomment-1173590862.

**Review by jeswr — COMMENTED** [review]
"Approved" @RubenVerborgh  with one suggested change (I don't have permissions to click the actual review button :))

**Comment by RubenVerborgh** [comment]
> I usually call `this._source.read()` outside of `if` statements as this reduces the amount of branching

Yeah; and I tend to optimize for fewer `return` points. The performance should not be impacted (and thanks for confirming that).

> Ooh interesting - I thought those were failing before without.
> If tests are working then I think it is safe to get rid of it provided we have good enough coverage with asynchronous sources for the iterator.

Yes—and that's the benefit of not messing with the base `readable` mechanism. It contains several guards and is ridiculously well tested for all kinds of orders. So we use the same logic now and inherit the same protections.

> "Approved" @RubenVerborgh with one suggested change (I don't have permissions to click the actual review button :))

And fully agreed—the code is even leaner now. I really like what we landed on:

```typescript
  read(): D | null {
    if (!this.done) {
      // Try to read an item that maps to a non-null value
      if (this._source.readable) {
        let item: S | null, mapped: D | null;
        while ((item = this._source.read()) !== null) {
          if ((mapped = this._map(item)) !== null)
            return mapped;
        }
      }
      this.readable = false;

      // Close this iterator if the source is empty
      if (this._source.done)
        this.close();
    }
    return null;
  }
```

This has been a common occurrence for me with `AsyncIterator`: lots of trying and fiddling, and eventually, you end up with really simple code. However, all of the fiddling resulted in lots of tests for edge cases that you wouldn't have found even if you would have thought of the simple code at the start (and you'd probably have gotten 1 or 2 lines wrong, but wouldn't know).

File: asynciterator.ts [hunk 14]
```diff
@@ -469,7 +474,9 @@ export class AsyncIterator<T> extends EventEmitter {
   filter<K extends T>(filter: (item: T) => item is K, self?: any): AsyncIterator<K>;
   filter(filter: (item: T) => boolean, self?: any): AsyncIterator<T>;
   filter(filter: (item: T) => boolean, self?: any): AsyncIterator<T> {
-    return this.transform({ filter: self ? filter.bind(self) : filter });
+    return this.map(function (this: any, item: T) {
+      return filter.call(self || this, item) ? item : null;
+    });
```
Comment (RubenVerborgh) [inline asynciterator.ts:479]: @jeswr Found this nasty bug still when restoring all of the main tests. (The `this` pointer `filter` is called with would be wrong because of the nested function.)

**Comment by RubenVerborgh** [comment]
Thanks a lot, @jeswr, excellent work. Thanks @jacoscaz as well.

**Comment by jacoscaz** [comment]
I had very little to do with this! Great job. I'll rebase #63 ASAP.

**Comment by RubenVerborgh** [comment]
@jacoscaz Reviews and chats are important! See Gitter 😉

**Comment by rubensworks** [comment]
Nice work!

@jeswr Do you have any insights into the performance impact on Comunica?
